include/sisl/status_mgr/status_mgr.hpp

```diff
@@ -0,0 +1,103 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <vector>
+#include <nlohmann/json.hpp>
+
+#include <sisl/logging/logging.h>
+
+namespace sisl {
+
+struct status_request {
+    nlohmann::json json;
+    bool do_recurse{false};
+    int verbose_level = 0;
+    std::vector< std::string > id_vec;
+};
+
+struct status_response {
+    nlohmann::json json;
+};
+
+using get_callback_type = std::function< status_response(const status_request&) >;
+class status_object;
+using status_object_ptr = std::shared_ptr< status_object >;
+
+// Keeps a heirarchy of modules/subsystems which register their callbacks to be called
+// whenever a get status is called from the root or directly.
+class status_object {
+    friend class status_manager;
+
+public:
+    status_object(const std::string& name, const std::string& type, get_callback_type cb) :
+            m_name(name), m_type(type), m_get_status_cb(std::move(cb)) {}
```

Comment:
`cb` is not taken as an rvalue reference so this is just moving the copy. I think taking callbacks by copy is fine right? It's relatively cheap and happens only at startup.

Reply:
Yeah right